tables/schema.py

```diff
@@ -0,0 +1,281 @@
+        return list(range(self.num_header_rows))  # assumes header starts from first row and is contiguous
+
+    @property
+    def table_width(self):
+        return self.num_dropped_columns + len(self.columns)
+
+    def extract(self, worksheet: pd.DataFrame):
+        """Extracts and processes all tables from the given worksheet that match the Schemas ID tag."""
+        extracted_tables, paired_tags = self._extract_tables_by_id(worksheet, self.id_tag, self.table_width)
+        processed_tables = self._process_tables(extracted_tables, paired_tags)
+        return processed_tables
+
+    def validate(self, table: pd.DataFrame) -> tuple[pd.DataFrame | None, list[ErrorMessage] | None]:
+        """Validates a table against the schema using pandera.
+
+        :param table: table to validate
+        :return: validated tables, error messages
+        """
+        validated_table = None
+        error_messages = None
+
+        try:
+            # TODO: currently we won't return tables that don't pass validation, which means any later validation
+            #   i.e. specific_validation wouldn't be executed, leading to a suboptimal "2 stage" validation process
+            #   Can we return the extracted (unvalidated, and therefore un-coerced tables?
+            #   Or should this be left for the calling function to deal with
+            validated_table = self._pandera_schema.validate(table, lazy=True)
+        except pa.errors.SchemaErrors as schema_errors:
+            error_messages = self._handle_errors(schema_errors, table)
+
+        return validated_table, error_messages
+
+    def _handle_errors(self, schema_errors: pa.errors.SchemaErrors, table: pd.DataFrame):
+        """Handle SchemaErrors by converting them to messages that refer to the original spreadsheet.
+
+        :param schema_errors: schema errors found during validation
+        :return: error messages
+        """
+        if not hasattr(table, "header_to_letter"):
+            raise TableExtractError(
+                "Table does not contain header to letter mapping - ensure that the table has not "
+                "been operated on between extract and validate."
+            )
+
+        error_messages = []
+        for failure in schema_errors.failure_cases.itertuples(name="Failure", index=False):
+            error_type = failure.check.split("(")[0]
+
+            # "dtype" check is series-wise and so will not produce element-wise messages. The solution to
+            # this is to use "coerce_dtype" errors instead as these are raised per element.
+            if error_type == "dtype":
+                continue
+
+            if error_type == "column_in_schema":
+                # "column_in_schema" means that the table didn't fit the schema - dev error rather than user
+                # TODO: should we use a different exception?
+                raise TableExtractError("Validated table contains columns from outside of the schema")
```

Comment:
```TableExtractError``` sounds correct for this use case. Is there a way of including these extraneous column names in the message or error?

Reply:
yeah good shout